lib/Modules/ComputeHooks/ComputeHooks.psm1

```diff
@@ -106,1360 +56,800 @@ function Install-Prerequisites {
-        [object]$Info
-    )
-    $exists = Get-Netadapter $juju_br
-    if(!$exists) {
-        Throw "Could not find OVS adapter"
-    }
-    $ips = $Info["addresses"]
-    if(!$ips -or !$ips.Count) {
-        return
+    $params = @{
+        'Namespace' = "novahyperv"
+        'Key' = "vmswitch_name"
     }
-    foreach ($i in $ips) {
-        $hasAddr = Get-NetIPAddress -AddressFamily $i["AddressFamily"].Trim() -IPAddress $i["IPAddress"].Trim() -PrefixLength $i["PrefixLength"] -ErrorAction SilentlyContinue
-        if($hasAddr) {
-            if($hasAddr.InterfaceIndex -ne $exists.ifIndex) {
-                $hasAddr | Remove-NetIPAddress -Confirm:$false | Out-Null
-            } else {
-                continue
-            }
-        }
-        if ($i["AddressFamily"] -eq "IPv6") { continue }
-        New-NetIPAddress -IPAddress $i["IPAddress"].Trim() -PrefixLength $i["PrefixLength"] -InterfaceIndex $exists.ifIndex| Out-Null
+    $vmSwitchName = Get-CharmState @params
+    if($vmSwitchName) {
+        Remove-VMSwitch -Name $vmSwitchName -Force -Confirm:$false
+        Remove-CharmState @params
     }
-    return
 }
 
-function Ensure-InternalOVSInterfaces {
-    $dataPort = Get-DataPortFromDataNetwork
-    if(!$dataPort) {
-        Throw "Failed to find data port"
-    }
-    $adapterInfo = Get-CharmState -Namespace "novahyperv" -Key "adapter_info"
-
-    Invoke-JujuCommand -Command @($ovs_vsctl, "--may-exist", "add-br", $juju_br)
-    Invoke-JujuCommand -Command @($ovs_vsctl, "--may-exist", "add-port", $juju_br, $adapterInfo["name"])
-
-    # Enable the OVS tap device
-    Get-Netadapter $juju_br | Enable-NetAdapter
-    Ensure-OVSTapDeviceAddress -Info $adapterInfo
-}
-
-function Install-OVS {
-    param(
-        [Parameter(Mandatory=$true)]
-        [string]$InstallerPath
-    )
-
-    Write-JujuInfo "Running OVS install"
-    $ovs = Get-ManagementObject -Class Win32_Product | Where-Object {$_.Name -match "open vswitch"}
-    if ($ovs){
-        Write-JujuInfo "OVS is already installed"
-        return $true
-    }
-
-    $hasInstaller = Test-Path $InstallerPath
-    if($hasInstaller -eq $false){
-        $InstallerPath = Get-OVSInstaller
-    }
-    Write-JujuInfo "Installing from $InstallerPath"
-    $ret = Start-Process -FilePath msiexec.exe -ArgumentList "INSTALLDIR=`"$ovsInstallDir`"","/qb","/l*v","$env:APPDATA\ovs-log.txt","/i","$InstallerPath" -Wait -PassThru
-    if($ret.ExitCode) {
-        Throw "Failed to install OVS: $LASTEXITCODE"
-    }
-    return $true
-}
+function Start-ConfigChangedHook {
+    Start-UpgradeOpenStackVersion
 
-function Check-OVSPrerequisites {
-    try {
-        $ovsdbSvc = Get-Service "ovsdb-server"
-        $ovsSwitchSvc = Get-Service "ovs-vswitchd"
-    } catch {
-        $InstallerPath = Get-OVSInstaller
-        Install-OVS $InstallerPath
-    }
-    if(!(Test-Path $ovs_vsctl)){
-        Throw "Could not find ovs-vsctl.exe in location: $ovs_vsctl"
-    }
+    Confirm-CharmPrerequisites
```

Comment:
Feels like Prerequisites should precede any other operation (including upgrade).